### ISSUE TITLE
LogSoftmax + NLLLoss not working, CrossEntropyLoss working

### ISSUE THREAD

**Issue by dennymarcels** [skorch-dev/skorch#519]
Hi!

I am trying to solve a classification problem with more than one class. I set up a network with some linear layers and a final LogSoftmax layer, then I chose the criterion to be NLLLoss, and I passed them as parameters to NeuralNetClassifier, to be used with cross_val_score. The training loss was returning as nan. When I investigated further, I noticed that the NLLLoss function was receiving nan as the predicted values (input parameter), but the classifier was actually returning valid values - only, since they are log of probabilities, all were negative. I believe NLLLoss is receiving a new log of the predicted outputs, instead of the raw outputs, therefore nans. Curiously, if I remove the LogSoftmax layer and use CrossEntropyLoss instead (which combines LogSoftmax and NLLLoss), the losses are calculated correctly. I could go no further in my investigation. I would appreciate some return. 

Thank you!

**Comment by BenjaminBossan** [comment]
> I am trying to solve a classification problem with more than one class

Well I sure hope there is more than one class ;)

> the classifier was actually returning valid values - only, since they are log of probabilities, all were negative

When using `NLLLoss`, please use `Softmax` as the output nonlinearity.

**Comment by dennymarcels** [comment]
> > I am trying to solve a classification problem with more than one class
> 
> Well I sure hope there is more than one class ;)
Haha well played.
> > the classifier was actually returning valid values - only, since they are log of probabilities, all were negative
> 
> When using `NLLLoss`, please use `Softmax` as the output nonlinearity.
Oh it did work, thank you! But can you please explain me why when training outside of Skorch (meaning, Pytorch alone) I should choose LogSoftmax (otherwise losses will be negative and diverge from 0), but then using Skorch to wrap it for Sklearn I should choose Softmax? That is even more intriguing to me once CrossEntropyLoss worked.

**Comment by BenjaminBossan** [comment]
> I noticed that my precision will be around 75%, while when using CrossEntropyLoss (and skipping Softmax altogether), they get up to 98%

That is interesting. Except from some numerical stability issues, the two approaches should yield the same result. If you could provide a minimum working example, I could investigate further.

**Comment by dennymarcels** [comment]
Sure, that would be great! The zipped db is attached.

Curiously, if I try fit on classifier_sklearn, the accuracy is also around 98%...

```
[iris.zip](https://github.com/skorch-dev/skorch/files/3562499/iris.zip)


import pandas as pd
import torch.nn as nn        
from skorch import NeuralNetClassifier
import torch
from sklearn.model_selection import cross_val_score

df = pd.read_csv('iris.csv')
X= df.iloc[:, 0:4].values
y= df.iloc[:, 4].values

from sklearn.preprocessing import LabelEncoder
labelencoder = LabelEncoder()
y = labelencoder.fit_transform(y)

X = X.astype('float32')
y = y.astype('int64')

class classifier_torch(nn.Module):
    def __init__(self):
        super().__init__()
        self.dense0 = nn.Linear(4, 4)
        self.dense1 = nn.Linear(4, 4)
        self.dense2 = nn.Linear(4, 3)
        self.activation = nn.ReLU()
        self.output = nn.Softmax()
        
    def forward(self, X):
        X = self.dense0(X)
        X = self.activation(X)
        X = self.dense1(X)
        X = self.activation(X)
        X = self.dense2(X)
        X = self.output(X)
        return X

classifier_sklearn = NeuralNetClassifier(module = classifier_torch, 
                                         criterion = torch.nn.NLLLoss, 
                                         optimizer = torch.optim.Adam,
                                         max_epochs = 100,
                                         batch_size = 10,
                                         train_split = False)

results = cross_val_score(classifier_sklearn, X, y, cv = 5, scoring = 'accuracy')
```

**Comment by dennymarcels** [comment]
I am also curious as to why use Softmax instead of LogSoftmax, as the [documentation](https://pytorch.org/docs/stable/_modules/torch/nn/modules/loss.html) clearly states:

> The `input` given through a forward call is expected to contain log-probabilities of each class.

**Comment by BenjaminBossan** [comment]
I ran your code with a couple of fixes (e.g. random seed 0) with both conditions, softmax + nllloss and logsoftmax + crossentropyloss. I didn't find any significant difference:

```python
>>> result  # sofmtax
array([1.        , 0.33333333, 1.        , 0.96666667, 0.9       ])
>>> result # crossentropy
array([1.        , 0.66666667, 1.        , 1.        , 0.93333333])
```

Yes, the mean is higher for entropy, but as you can see from the individual scores, there is a lot of variance (other random seeds do reverse the results). Therefore, I am sure that the two approaches are equivalent.

Next comes the question why there is so much variance in the data.

At the beginning, there was even more variance. The reason is, you didn't shuffle the data, but the target is sorted by default (`[0, 0, 0, ..., 1, 1, 1, ..., 2, 2, 2]`), which can lead to very imbalanced splits with sklearn's cv splits. I did shuffle the data, which already decreased variance. Why is there still so much variance left? Probably the amount of data -- 150 samples -- is just too small to avoid it, but I didn't investigate further.

> I am also curious as to why use Softmax instead of LogSoftmax

This is mainly for historical reasons. At the time we published skorch, libraries like keras and nolearn were widely used (keras still is). There, the default for classifiers was to return probabilities, hence softmax is needed. We wanted to be consistent. Also, this way, your `net.predict_proba` can return the module output without any further transformation, making this step more transparent.

If we had to redesign skorch from sketch, we might make a different decision now, but at this point, it would break far too much code.

**Comment by dennymarcels** [comment]
Thanks a lot for the effort, I understand your approach and agree that the issues I am concerned with might be due to the problems you point. One last question: you said you used logsoftmax + crossentropyloss, but with crossentropyloss we are not supposed to used a (non-linear transformation) layer at the end of the net. Did you really include it? If so, why would that be the case?

**Comment by BenjaminBossan** [comment]
> you said you used logsoftmax + crossentropyloss, but with crossentropyloss we are not supposed to used a (non-linear transformation) layer at the end of the net

I tested without explicit logsoftmax. But I think it would work either way.

If you have no further questions, feel free to close the issue.

**Comment by dennymarcels** [comment]
Alrighty, thank you!
